### PR TITLE
Close translog view after primary-replica resync

### PR DESCRIPTION
**Pull request by ywelsch**
The translog view was being closed too early, possibly causing a failed resync. Note: The bug only affects unreleased code.

Relates to #24841.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java [hunk 1]
```diff
@@ -96,16 +118,20 @@ public synchronized int totalOperations() {
...
                 @Override
                 public synchronized Translog.Operation next() throws IOException {
-                    if (indexShard.state() != IndexShardState.STARTED) {
-                        assert indexShard.state() != IndexShardState.RELOCATED : "resync should never happen on a relocated shard";
-                        throw new IndexShardNotStartedException(shardId, indexShard.state());
+                    IndexShardState state = indexShard.state();
+                    if (state == IndexShardState.CLOSED) {
+                        throw new IndexShardClosedException(shardId);
+                    } else {
+                        assert state == IndexShardState.STARTED : "resync should only happen on a started shard";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: ++ . nit: add the state to the message please.

File: core/src/test/java/org/elasticsearch/index/shard/PrimaryReplicaSyncerTests.java [hunk 2]
```diff
@@ -90,6 +91,47 @@ public void testSyncerSendsOffCorrectDocuments() throws Exception {
...
+        PlainActionFuture<PrimaryReplicaSyncer.ResyncTask> fut = new PlainActionFuture<>();
+        threadPool.generic().execute(() -> {
+            try {
+                syncer.resync(shard, fut);
+            } catch (AlreadyClosedException ace) {
+                fut.onFailure(ace);
+            }
+        });
+        if (randomBoolean()) {
+            assertBusy(() -> assertTrue("Sync action was not called", syncActionCalled.get()));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/PrimaryReplicaSyncerTests.java]: nit: this pains me :) can we use a latch?

**Review by bleskes — APPROVED** [review]
Thanks @ywelsch

**Comment by ywelsch** [comment]
Thanks @bleskes

**Pull request merged**